File: python/src/linksql/substitution.py
"""The single substitution operation.

LinksQL has exactly one operation. A query is a pair of pattern lists — a
**restriction** and a **substitution** — written ``(restriction) (substitution)``.
The restriction selects existing links (binding variables along the way); the
substitution describes what those links become. The four CRUD behaviours are not
separate primitives, they are derived from how the two lists line up:

==========  ==============  ==========================================
behaviour   form            meaning
==========  ==============  ==========================================
create      ``()  ((s t))`` empty restriction, one created link
read        ``((p))``       restriction only, no mutation
update      ``((a b)) ((a c))`` matched link rewritten in place
delete      ``((a b)) ()``  matched link with no replacement is removed
==========  ==============  ==========================================

Patterns are paired positionally: ``restriction[i]`` becomes ``substitution[i]``.
Any trailing restriction patterns are deletions, any trailing substitution
patterns are creations. This positional pairing, combined with variables, makes
the operation a Markov algorithm over the link space — and therefore Turing
complete.

This module is a faithful, behaviourally-identical port of
``src/substitution.js`` from the JavaScript reference implementation.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from .lino import Link as LinkNode
from .lino import Node, Ref

if TYPE_CHECKING:
    from .names import Names
    from .store import Link, LinksStore


class SubstitutionError(Exception):
    """Error raised when a substitution cannot be carried out."""


@dataclass
class Context:
    """Execution context shared by the matching and substitution passes.

    ``names`` is optional: when ``None`` named references cannot be resolved
    (matching yields "unresolved", output resolution is an error), mirroring the
    ``ctx.names ?`` guards in the JavaScript reference.
    """

    store: LinksStore
    names: Names | None = None


@dataclass(frozen=True)
class Slots:
    """The three slots of a decomposed link pattern: ``(id, source, target)``.

    A ``None`` slot means "unconstrained" when matching and "inherit from the
    matched link" when producing an update.
    """

    id: Node | None
    source: Node | None
    target: Node | None


@dataclass
class Row:
    """A join row: the concrete links matched and the variable bindings."""

    binding: dict[str, int]
    links: list[Link]


@dataclass
class Spec:
    """A concrete link specification produced by materialising a pattern."""

    index: int | None
    source: int
    target: int


@dataclass
class RawResult:
    """The internal result of :func:`execute`, before report formatting.

    Mirrors the ``{matches, created, updated, deleted}`` object returned by the
    JavaScript ``execute``; exposed as a dataclass so callers use attribute
    access and mypy can check every field.
    """

    matches: list[Row] = field(default_factory=list)
    created: list[Link] = field(default_factory=list)
    updated: list[Link] = field(default_factory=list)
    deleted: list[Link] = field(default_factory=list)


def link_slots(node: Node) -> Slots:
    """Decompose a link pattern node into its three slots.

    A doublet pattern is ``(id: source target)``. The identity slot is optional,
    and a pattern may also be written ``()`` (match anything) or ``(id)`` (match
    a single identity, any structure).
    """
    if not isinstance(node, LinkNode):
        message = "Each restriction/substitution pattern must be a link"
        raise SubstitutionError(message)
    values = node.values
    if len(values) == 2:
        return Slots(id=node.id, source=values[0], target=values[1])
    if len(values) == 0:
        return Slots(id=node.id, source=None, target=None)
    if len(values) == 1 and node.id is None:
        return Slots(id=values[0], source=None, target=None)
    message = f"A link pattern must have 0 or 2 values (got {len(values)})"
    raise SubstitutionError(message)


def _resolve_for_match(
    node: Node | None, binding: dict[str, int], ctx: Context
) -> int | None:
    """Resolve a value node to a concrete link index for *matching* purposes.

    Returns ``None`` when the node cannot resolve.
    """
    if isinstance(node, Ref):
        if node.kind == "number":
            assert isinstance(node.value, int)
            return node.value
        if node.kind == "name":
            assert isinstance(node.value, str)
            return ctx.names.resolve(node.value) if ctx.names is not None else None
        # variables and wildcards are handled by the caller
        return None
    if node is None:
        return None
    slots = link_slots(node)
    source = _resolve_for_match(slots.source, binding, ctx)
    target = _resolve_for_match(slots.target, binding, ctx)
    if source is None or target is None:
        return None
    found = ctx.store.find_by_pair(source, target)
    return found.index if found is not None else None


def _match_slot(
    slot: Node | None, actual: int, binding: dict[str, int], ctx: Context
) -> bool:
    """Constrain one slot of a pattern against an actual value.

    Updates ``binding`` in place when a fresh variable is bound; ``None`` slots
    are unconstrained.
    """
    if slot is None:
        return True
    if isinstance(slot, Ref):
        if slot.kind == "wildcard":
            return True
        if slot.kind == "variable":
            assert isinstance(slot.value, str)
            if slot.value in binding:
                return binding[slot.value] == actual
            binding[slot.value] = actual
            return True
    expected = _resolve_for_match(slot, binding, ctx)
    return expected is not None and expected == actual


def _match_one(
    pattern: Node, link: Link, binding: dict[str, int], ctx: Context
) -> bool:
    """Test a single pattern against a single link, extending ``binding``."""
    slots = link_slots(pattern)
    return (
        _match_slot(slots.id, link.index, binding, ctx)
        and _match_slot(slots.source, link.source, binding, ctx)
        and _match_slot(slots.target, link.target, binding, ctx)
    )


def _join_restriction(
    patterns: list[Node], ctx: Context, snapshot: list[Link]
) -> list[Row]:
    """Join all restriction patterns into a set of binding rows.

    Each row records the concrete link matched by every pattern, in order, so the
    substitution can pair with them positionally.
    """
    rows: list[Row] = [Row(binding={}, links=[])]
    for pattern in patterns:
        nxt: list[Row] = []
        for row in rows:
            for link in snapshot:
                binding = dict(row.binding)
                if _match_one(pattern, link, binding, ctx):
                    nxt.append(Row(binding=binding, links=[*row.links, link]))
        rows = nxt
    return rows


def _resolve_for_output(
    node: Node, binding: dict[str, int], ctx: Context, created: list[Link]
) -> int:
    """Resolve a value node to a concrete index when *producing* output.

    Unlike matching, every reference must resolve: unbound variables and
    wildcards are errors, and names are auto-created when the context allows it.
    Links created as a side effect are appended to ``created``.
    """
    if isinstance(node, Ref):
        if node.kind == "number":
            assert isinstance(node.value, int)
            return node.value
        if node.kind == "variable":
            assert isinstance(node.value, str)
            if node.value not in binding:
                message = f"Unbound variable ${node.value} in substitution"
                raise SubstitutionError(message)
            return binding[node.value]
        if node.kind == "wildcard":
            message = "Wildcard * cannot appear in a substitution"
            raise SubstitutionError(message)
        assert isinstance(node.value, str)
        if ctx.names is None:
            message = f'Named reference "{node.value}" requires names to be enabled'
            raise SubstitutionError(message)
        return ctx.names.ensure(node.value)
    slots = link_slots(node)
    if slots.source is None or slots.target is None:
        message = "Nested link must have a source and a target"
        raise SubstitutionError(message)
    source = _resolve_for_output(slots.source, binding, ctx, created)
    target = _resolve_for_output(slots.target, binding, ctx, created)
    index = (
        None
        if slots.id is None
        else _resolve_for_output(slots.id, binding, ctx, created)
    )
    existing = ctx.store.find_by_pair(source, target)
    if existing is not None:
        return existing.index
    link = ctx.store.create(index=index, source=source, target=target)
    created.append(link)
    return link.index


def _materialize(
    pattern: Node,
    binding: dict[str, int],
    ctx: Context,
    created: list[Link],
    matched: Link | None = None,
) -> Spec:
    """Turn a substitution pattern into a concrete ``Spec``.

    ``matched`` is the link being rewritten (for an update); when both
    source/target slots are ``None`` the spec inherits its structure.
    """
    slots = link_slots(pattern)
    index = (
        None
        if slots.id is None
        else _resolve_for_output(slots.id, binding, ctx, created)
    )
    if slots.source is None or slots.target is None:
        if matched is None:
            message = "A created link must specify a source and a target"
            raise SubstitutionError(message)
        return Spec(index=index, source=matched.source, target=matched.target)
    source = _resolve_for_output(slots.source, binding, ctx, created)
    target = _resolve_for_output(slots.target, binding, ctx, created)
    return Spec(index=index, source=source, target=target)


def match_restriction(restriction: list[Node], ctx: Context) -> list[Row]:
    """Match a restriction against the store without mutating anything.

    This is the read path: a query with a restriction but no substitution returns
    its matches verbatim. Keeping it separate from :func:`execute` is what lets a
    lone restriction mean "read" rather than "delete".
    """
    return _join_restriction(restriction, ctx, ctx.store.all())


def link_matches(patterns: list[Node], link: Link, ctx: Context) -> bool:
    """Test whether a link satisfies at least one of the given patterns.

    Each pattern is tried with a fresh binding. An empty pattern list always
    matches.
    """
    if len(patterns) == 0:
        return True
    return any(_match_one(pattern, link, {}, ctx) for pattern in patterns)


def execute(
    restriction: list[Node], substitution: list[Node], ctx: Context
) -> RawResult:
    """Execute one substitution operation against the store."""
    snapshot = ctx.store.all()
    rows = _join_restriction(restriction, ctx, snapshot)
    result = RawResult()
    paired = min(len(restriction), len(substitution))

    for row in rows:
        result.matches.append(Row(binding=row.binding, links=row.links))

        for i in range(paired):
            matched = row.links[i]
            if not ctx.store.has(matched.index):
                continue  # already removed by an earlier row
            spec = _materialize(
                substitution[i], row.binding, ctx, result.created, matched
            )
            result.updated.append(
                ctx.store.update(
                    matched.index,
                    source=spec.source,
                    target=spec.target,
                    new_index=spec.index,
                )
            )

        for i in range(paired, len(restriction)):
            matched = row.links[i]
            if ctx.store.delete(matched.index):
                result.deleted.append(matched)

        for i in range(paired, len(substitution)):
            spec = _materialize(substitution[i], row.binding, ctx, result.created)
            result.created.append(
                ctx.store.create(
                    index=spec.index, source=spec.source, target=spec.target
                )
            )

    return result

File: csharp/src/LinksQL/Substitution.cs
namespace LinksQL;

/// <summary>
/// The execution context passed to the substitution engine: the backing store
/// and an optional name registry.
/// </summary>
/// <param name="Store">The links store to read and mutate.</param>
/// <param name="Names">
/// The name registry, or <see langword="null"/> when named references are
/// disabled.
/// </param>
public sealed record SubstitutionContext(LinksStore Store, Names? Names);

/// <summary>
/// The three slots a link pattern decomposes into.
/// </summary>
/// <param name="Id">The identity slot, or <see langword="null"/> when absent.</param>
/// <param name="Source">The source slot, or <see langword="null"/> ("match anything").</param>
/// <param name="Target">The target slot, or <see langword="null"/> ("match anything").</param>
public readonly record struct LinkSlots(Node? Id, Node? Source, Node? Target);

/// <summary>
/// A single binding row produced by the restriction join: the concrete links
/// matched (in pattern order) and the variable bindings shared across the row.
/// </summary>
public sealed class Row
{
    /// <summary>Gets the variable bindings as <c>name -&gt; index</c>.</summary>
    /// <remarks>
    /// Mutable during the join (each pattern extends the binding in place);
    /// stable once the row is finalised.
    /// </remarks>
    public Dictionary<string, long> Binding { get; init; } = new(StringComparer.Ordinal);

    /// <summary>Gets the concrete links matched, in pattern order.</summary>
    public IReadOnlyList<Link> Links { get; init; } = Array.Empty<Link>();
}

/// <summary>
/// A report of everything an <see cref="Substitution.Execute"/> matched and
/// changed.
/// </summary>
/// <remarks>
/// The engine appends to the <c>internal</c> backing lists while running; the
/// public surface exposes only read-only views.
/// </remarks>
public sealed class ExecutionResult
{
    internal List<Row> MatchesList { get; } = new();

    internal List<Link> CreatedList { get; } = new();

    internal List<Link> UpdatedList { get; } = new();

    internal List<Link> DeletedList { get; } = new();

    /// <summary>Gets the join rows that matched the restriction.</summary>
    public IReadOnlyList<Row> Matches => MatchesList;

    /// <summary>Gets the links created by this operation.</summary>
    public IReadOnlyList<Link> Created => CreatedList;

    /// <summary>Gets the links updated by this operation (final structure).</summary>
    public IReadOnlyList<Link> Updated => UpdatedList;

    /// <summary>Gets the links removed by this operation.</summary>
    public IReadOnlyList<Link> Deleted => DeletedList;
}

/// <summary>
/// The single substitution operation and its matching primitives.
/// </summary>
/// <remarks>
/// LinksQL has exactly one operation. A query is a pair of pattern lists — a
/// <em>restriction</em> and a <em>substitution</em>. The restriction selects
/// existing links (binding variables along the way); the substitution describes
/// what those links become. The four CRUD behaviours are derived from how the
/// two lists line up positionally.
/// </remarks>
public static class Substitution
{
    /// <summary>
    /// Decompose a link pattern node into its three slots.
    /// </summary>
    /// <param name="node">A <see cref="LinkNode"/>.</param>
    /// <returns>The decomposed <see cref="LinkSlots"/>.</returns>
    /// <exception cref="SubstitutionError">
    /// Thrown when the node is not a link, or has an unsupported value count.
    /// </exception>
    public static LinkSlots GetLinkSlots(Node node)
    {
        if (node is not LinkNode link)
        {
            throw new SubstitutionError(
                "Each restriction/substitution pattern must be a link");
        }

        var values = link.Values;
        if (values.Count == 2)
        {
            return new LinkSlots(link.Id, values[0], values[1]);
        }

        if (values.Count == 0)
        {
            return new LinkSlots(link.Id, null, null);
        }

        if (values.Count == 1 && link.Id is null)
        {
            return new LinkSlots(values[0], null, null);
        }

        throw new SubstitutionError(
            $"A link pattern must have 0 or 2 values (got {values.Count})");
    }

    /// <summary>
    /// Resolve a value node to a concrete link index for <em>matching</em>
    /// purposes.
    /// </summary>
    /// <param name="node">Reference or nested link node.</param>
    /// <param name="binding">Current variable bindings.</param>
    /// <param name="ctx">Execution context.</param>
    /// <returns>The index, or <see langword="null"/> when it cannot resolve.</returns>
    private static long? ResolveForMatch(Node? node, Dictionary<string, long> binding, SubstitutionContext ctx)
    {
        switch (node)
        {
            case null:
                // Mirrors the JavaScript engine, which never reaches here with a
                // null slot in practice; treat "no node" as unresolved.
                return null;
            case NumberRef number:
                return number.Value;
            case NameRef name:
                return ctx.Names?.Resolve(name.Value);
            case VariableRef:
            case WildcardRef:
                // Variables and wildcards are handled by the caller.
                return null;
        }

        var slots = GetLinkSlots(node);
        var source = ResolveForMatch(slots.Source, binding, ctx);
        var target = ResolveForMatch(slots.Target, binding, ctx);
        if (source is null || target is null)
        {
            return null;
        }

        var found = ctx.Store.FindByPair(source.Value, target.Value);
        return found?.Index;
    }

    /// <summary>
    /// Constrain one slot of a pattern against an actual value, updating bindings.
    /// </summary>
    /// <param name="slot">The slot node (<see langword="null"/> means "match anything").</param>
    /// <param name="actual">The link's value at this slot.</param>
    /// <param name="binding">Mutable bindings for the current row.</param>
    /// <param name="ctx">Execution context.</param>
    /// <returns><see langword="true"/> when the slot is satisfied.</returns>
    public static bool MatchSlot(Node? slot, long actual, Dictionary<string, long> binding, SubstitutionContext ctx)
    {
        ArgumentNullException.ThrowIfNull(binding);
        ArgumentNullException.ThrowIfNull(ctx);

        switch (slot)
        {
            case null:
                return true;
            case WildcardRef:
                return true;
            case VariableRef variable:
                if (binding.TryGetValue(variable.Value, out var bound))
                {
                    return bound == actual;
                }

                binding[variable.Value] = actual;
                return true;
        }

        var expected = ResolveForMatch(slot, binding, ctx);
        return expected is not null && expected.Value == actual;
    }

    /// <summary>
    /// Test a single pattern against a single link, extending the binding in
    /// place.
    /// </summary>
    /// <param name="pattern">Link pattern node.</param>
    /// <param name="link">A stored link.</param>
    /// <param name="binding">Mutable bindings.</param>
    /// <param name="ctx">Execution context.</param>
    /// <returns><see langword="true"/> when the link matches the pattern.</returns>
    public static bool MatchOne(Node pattern, Link link, Dictionary<string, long> binding, SubstitutionContext ctx)
    {
        ArgumentNullException.ThrowIfNull(link);

        var slots = GetLinkSlots(pattern);
        return MatchSlot(slots.Id, link.Index, binding, ctx)
            && MatchSlot(slots.Source, link.Source, binding, ctx)
            && MatchSlot(slots.Target, link.Target, binding, ctx);
    }

    /// <summary>
    /// Join all restriction patterns into a set of binding rows.
    /// </summary>
    /// <param name="patterns">Restriction pattern nodes.</param>
    /// <param name="ctx">Execution context.</param>
    /// <param name="snapshot">The links to match against (pre-mutation).</param>
    /// <returns>The surviving binding rows.</returns>
    public static IReadOnlyList<Row> JoinRestriction(
        IReadOnlyList<Node> patterns,
        SubstitutionContext ctx,
        IReadOnlyList<Link> snapshot)
    {
        ArgumentNullException.ThrowIfNull(patterns);
        ArgumentNullException.ThrowIfNull(snapshot);

        var rows = new List<Row> { new() };
        foreach (var pattern in patterns)
        {
            var next = new List<Row>();
            foreach (var row in rows)
            {
                foreach (var link in snapshot)
                {
                    var binding = new Dictionary<string, long>(row.Binding, StringComparer.Ordinal);
                    if (MatchOne(pattern, link, binding, ctx))
                    {
                        var links = new List<Link>(row.Links) { link };
                        next.Add(new Row { Binding = binding, Links = links });
                    }
                }
            }

            rows = next;
        }

        return rows;
    }

    /// <summary>
    /// Match a restriction against the store without mutating anything (the read
    /// path).
    /// </summary>
    /// <param name="restriction">Restriction pattern nodes.</param>
    /// <param name="ctx">Execution context.</param>
    /// <returns>The binding rows.</returns>
    public static IReadOnlyList<Row> Match(IReadOnlyList<Node> restriction, SubstitutionContext ctx)
    {
        ArgumentNullException.ThrowIfNull(ctx);

        return JoinRestriction(restriction, ctx, ctx.Store.All());
    }

    /// <summary>
    /// Test whether a single link satisfies at least one of the given patterns.
    /// </summary>
    /// <param name="patterns">Restriction pattern nodes.</param>
    /// <param name="link">A link.</param>
    /// <param name="ctx">Execution context.</param>
    /// <returns><see langword="true"/> when the link matches any pattern.</returns>
    public static bool LinkMatches(IReadOnlyList<Node> patterns, Link link, SubstitutionContext ctx)
    {
        ArgumentNullException.ThrowIfNull(patterns);

        if (patterns.Count == 0)
        {
            return true;
        }

        foreach (var pattern in patterns)
        {
            if (MatchOne(pattern, link, new Dictionary<string, long>(StringComparer.Ordinal), ctx))
            {
                return true;
            }
        }

        return false;
    }

    /// <summary>
    /// Resolve a value node to a concrete index when <em>producing</em> output.
    /// Unlike matching, every reference must resolve.
    /// </summary>
    /// <param name="node">Reference or nested link node.</param>
    /// <param name="binding">Variable bindings for the row.</param>
    /// <param name="ctx">Execution context.</param>
    /// <param name="created">Accumulator for links created as a side effect.</param>
    /// <returns>The resolved index.</returns>
    /// <exception cref="SubstitutionError">
    /// Thrown for unbound variables, wildcards, unresolvable names, or malformed
    /// nested links.
    /// </exception>
    private static long ResolveForOutput(Node node, Dictionary<string, long> binding, SubstitutionContext ctx, ICollection<Link> created)
    {
        switch (node)
        {
            case NumberRef number:
                return number.Value;
            case VariableRef variable:
                if (!binding.TryGetValue(variable.Value, out var bound))
                {
                    throw new SubstitutionError(
                        $"Unbound variable ${variable.Value} in substitution");
                }

                return bound;
            case WildcardRef:
                throw new SubstitutionError("Wildcard * cannot appear in a substitution");
            case NameRef name:
                if (ctx.Names is null)
                {
                    throw new SubstitutionError(
                        $"Named reference \"{name.Value}\" requires names to be enabled");
                }

                return ctx.Names.Ensure(name.Value);
        }

        var slots = GetLinkSlots(node);
        if (slots.Source is null || slots.Target is null)
        {
            throw new SubstitutionError("Nested link must have a source and a target");
        }

        var source = ResolveForOutput(slots.Source, binding, ctx, created);
        var target = ResolveForOutput(slots.Target, binding, ctx, created);
        long? index = slots.Id is null ? null : ResolveForOutput(slots.Id, binding, ctx, created);
        var existing = ctx.Store.FindByPair(source, target);
        if (existing is not null)
        {
            return existing.Index;
        }

        var link = ctx.Store.Create(index, source, target);
        created.Add(link);
        return link.Index;
    }

    /// <summary>
    /// Turn a substitution pattern into a concrete <c>(index?, source, target)</c>
    /// spec.
    /// </summary>
    /// <param name="pattern">Substitution pattern node.</param>
    /// <param name="binding">Variable bindings for the row.</param>
    /// <param name="ctx">Execution context.</param>
    /// <param name="created">Accumulator for nested-link side effects.</param>
    /// <param name="matched">The link being rewritten (for an update), if any.</param>
    /// <returns>The materialised <c>(index, source, target)</c> specification.</returns>
    /// <exception cref="SubstitutionError">
    /// Thrown when a created link omits its source and target.
    /// </exception>
    public static (long? Index, long Source, long Target) Materialize(
        Node pattern,
        Dictionary<string, long> binding,
        SubstitutionContext ctx,
        ICollection<Link> created,
        Link? matched)
    {
        ArgumentNullException.ThrowIfNull(binding);
        ArgumentNullException.ThrowIfNull(ctx);
        ArgumentNullException.ThrowIfNull(created);

        var slots = GetLinkSlots(pattern);
        long? index = slots.Id is null ? null : ResolveForOutput(slots.Id, binding, ctx, created);
        if (slots.Source is null || slots.Target is null)
        {
            if (matched is null)
            {
                throw new SubstitutionError(
                    "A created link must specify a source and a target");
            }

            return (index, matched.Source, matched.Target);
        }

        var source = ResolveForOutput(slots.Source, binding, ctx, created);
        var target = ResolveForOutput(slots.Target, binding, ctx, created);
        return (index, source, target);
    }

    /// <summary>
    /// Execute one substitution operation against the store.
    /// </summary>
    /// <param name="restriction">Restriction pattern nodes.</param>
    /// <param name="substitution">Substitution pattern nodes.</param>
    /// <param name="ctx">Execution context.</param>
    /// <returns>A report of everything that matched and changed.</returns>
    public static ExecutionResult Execute(
        IReadOnlyList<Node> restriction,
        IReadOnlyList<Node> substitution,
        SubstitutionContext ctx)
    {
        ArgumentNullException.ThrowIfNull(restriction);
        ArgumentNullException.ThrowIfNull(substitution);
        ArgumentNullException.ThrowIfNull(ctx);

        var snapshot = ctx.Store.All();
        var rows = JoinRestriction(restriction, ctx, snapshot);
        var result = new ExecutionResult();
        var paired = Math.Min(restriction.Count, substitution.Count);

        foreach (var row in rows)
        {
            result.MatchesList.Add(row);

            for (var i = 0; i < paired; i += 1)
            {
                var matched = row.Links[i];
                if (!ctx.Store.Has(matched.Index))
                {
                    continue; // already removed by an earlier row
                }

                var spec = Materialize(substitution[i], row.Binding, ctx, result.CreatedList, matched);
                result.UpdatedList.Add(
                    ctx.Store.Update(matched.Index, spec.Source, spec.Target, spec.Index));
            }

            for (var i = paired; i < restriction.Count; i += 1)
            {
                var matched = row.Links[i];
                if (ctx.Store.Delete(matched.Index))
                {
                    result.DeletedList.Add(matched);
                }
            }

            for (var i = paired; i < substitution.Count; i += 1)
            {
                var spec = Materialize(substitution[i], row.Binding, ctx, result.CreatedList, null);
                result.CreatedList.Add(ctx.Store.Create(spec.Index, spec.Source, spec.Target));
            }
        }

        return result;
    }
}
